core/src/main/java/org/geysermc/geyser/extension/GeyserExtensionClassLoader.java

```diff
@@ -76,22 +82,40 @@ protected Class<?> findClass(String name) throws ClassNotFoundException {
     }
 
     protected Class<?> findClass(String name, boolean checkGlobal) throws ClassNotFoundException {
+        // Don't let extensions load classes from Geyser or minecraft packages without a warning
         if (name.startsWith("org.geysermc.geyser.") || name.startsWith("net.minecraft.")) {
-            throw new ClassNotFoundException(name);
+            if (!this.description.allowInternalCodeReferences()) {
+                throw new ClassNotFoundException(name);
```

Comment:
I'd include the warning message used below (minus the "This can change at any time..." part) in this exception rather than just the name.

Reply:
Changed in https://github.com/GeyserMC/Geyser/pull/3946/commits/30ced6325d3186b9ea30a09a2b1abed761a13f2f